soroban-env-host/src/native_contract/token/cryptography.rs

```diff
@@ -63,10 +71,10 @@ fn check_account_auth(
             auth.account_id.clone().into(),
             sig.public_key.clone().into(),
         )?;
-        let signer_weight: u32 = signer_weight_rv.try_into()?;
-        // TODO: Check for overflow
+        // Clamp signature weight to be at most 255. This is consistent with
+        // classic tx signature weight computations in Core.
+        let signer_weight: u32 = min(signer_weight_rv.try_into()?, MAX_ACCOUNT_SIGNATURE_WEIGHT);
```

Comment:
The weights are retrieved from the signers in core, so they should never be greater than u8::MAX. We should error if this is somehow false instead of clamping.

Reply:
Is the same true for the number of signers? If core never tells us about more than 20 signers, this fn can never succeed if more than 20 signers are provided, therefore we also don't really need the length check?

Reply:
Yeah that's true. The signers vector in AccountEntry has a max size of 20.

Reply:
Yeah that's true.

Reply:
> The weights are retrieved from the signers in core, so they should never be greater than u8::MAX. We should error if this is somehow false instead of clamping.

For whatever weird reason (I suppose historical), the weights are `u32` in `AccountEntry` and seemingly nothing prevents them from being treater than 255. The thresholds are indeed `u8`. This is why core does clamping too when checking the signatures: https://github.com/stellar/stellar-core/blob/master/src/transactions/SignatureChecker.cpp#L85. I suppose we could move clamping to the weight getter instead and change weight to be `u8`; I think that's probably slightly preferrable to doing double-checks. Please let me know if that sounds good to you.

> therefore we also don't really need the length check?

I'm not sure understand this argument. The signature vector is unbounded here, unlike in classic transaction XDR. The reason we have this limit is the core's limit on *signers* count, but here we limit *signatures* count.

Reply:
Ah that's a good catch and looks like I'm wrong. Signers added before protocol 10 can have weights greater than 255 - https://github.com/stellar/stellar-core/blob/3f895f4d0120cdc921710b518fc6f27ed8ac2140/src/transactions/SetOptionsOpFrame.cpp#L298.

Reply:
Yeah clamping in the host sounds like a better idea.

Reply:
Leighs point is that we iterate over the signatures here, and if we have say 21 signatures, the verification of the 21st signature will always fail.

Reply:
Sure, this may when there are more signatures. However I see the following benefits of having an explicit check:

- Currently we don't return error for signers that don't belong to account and just return 0 weight, hence it's possible to have >20 signatures and still get valid result. Maybe we should change this though.
- Shortcut for malformed signatures (no need to verify everything before failing)
- It's clear from code that we are protected from overflow/overly large vectors etc. At least I'm feeling a bit safer when I see that we explicitly check unbounded vector sizes instead of relying on some more complex logic.

Reply:
> Currently we don't return error for signers that don't belong to account and just return 0 weight, hence it's possible to have >20 signatures and still get valid result. Maybe we should change this though.

This seems like a bug. Extra signatures should be considered an error for consistency with stellar-core too.

Reply:
True, the comment actually says that the fn traps. Returning an error now.